src/main/java/edu/berkeley/cs186/database/query/join/GHJOperator.java

```diff
@@ -108,10 +122,37 @@ private void buildAndProbe(Partition leftPartition, Partition rightPartition) {
                 "fit in B-2 pages of memory."
             );
         }
-        // TODO(proj3_part1): implement the building and probing stage
+        // TODO(proj3_part1): implement the building and probing stage Done
         // You shouldn't refer to any variable starting with "left" or "right"
         // here, use the "build" and "probe" variables we set up for you.
         // Check out how SHJOperator implements this function if you feel stuck.
+        Map<DataBox, List<Record>> hashTable = new HashMap<>();
+
+        // build 과정
+        for (Record buildRecord : buildRecords) {
+            DataBox buildJoinValue = buildRecord.getValue(buildColumnIndex);
+            if (!hashTable.containsKey(buildJoinValue)) {
+                hashTable.put(buildJoinValue, new ArrayList<>());
+            }
+            hashTable.get(buildJoinValue).add(buildRecord);
+        }
+
+        // probe 과정
+        for (Record probeRecord : probeRecords) {
+            DataBox probeJoinValue = probeRecord.getValue(probeColumnIndex);
+            if (!hashTable.containsKey(probeJoinValue)) {
+                continue;
+            }
+            for (Record record : hashTable.get(probeJoinValue)) {
+                Record joinedRecord;
+                if (probeFirst) {
+                    joinedRecord = probeRecord.concat(record);
+                } else {
+                    joinedRecord = record.concat(probeRecord);
+                }
```

Comment:
```suggestion
                joinedRecord = getJoinRecordBy(probeFirst);
                
                
           private Record getJoinRecordBy(boolean probreFirst){
               Record joinedRecord;
                if (probeFirst) {
                    joinedRecord = probeRecord.concat(record);
                } else {
                    joinedRecord = record.concat(probeRecord);
                }
                return joinedRecord;
            }
```

이런식으로 바꾸면 if문도 줄이고 가독성도 좋아지지 않을까요?